### PR TITLE
[Tree assume] Add `IsValid`

### PR DESCRIPTION
**Pull request by rtjoa**
Add `IsValid` to tree-assume. See #262.

---

PR stack:
1. #336
2. #337 
3. #340 
4. #338 
3. #339

### CODE REVIEW

File: tree_assume/src/optimizations/is_valid.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+(fail (check (ExprIsValid {not_made_valid})))
+    ",
+        num0 = int(0),
+        arg = arg(),
+        tup45 = parallel!(int(4), int(5)),
+    );
+    crate::egglog_test(
+        &build,
+        &check,
+        vec![],
```
Comment (oflatt) [inline tree_assume/src/optimizations/is_valid.rs:61]: Pass the loop in here to check that it runs

File: tree_assume/src/optimizations/is_valid.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    ESort::iter()
+        .map(|sort| "(relation *IsValid (*))".replace('*', sort.name()))
+        .chain(Constructor::iter().filter_map(rule_for_ctor))
+        .collect::<Vec<_>>()
+}
+#[cfg(test)]
+use crate::ast::*;
+#[cfg(test)]
+use crate::Value;
+
```
Comment (oflatt) [inline tree_assume/src/optimizations/is_valid.rs:35]: Might be good to have a test that shows that validity passes through the assume node but not its left child

**Review by oflatt — CHANGES_REQUESTED** [review]
Looks good, had a couple suggestions

File: tree_assume/src/optimizations/is_valid.rs (hunk 1 above)
Comment (rtjoa) [reply tree_assume/src/optimizations/is_valid.rs]: Updated

File: tree_assume/src/optimizations/is_valid.rs (hunk 2 above)
Comment (rtjoa) [reply tree_assume/src/optimizations/is_valid.rs:37]: Done

**Review by oflatt — APPROVED** [review]
great!

**Pull request merged**